pkg/rehearse/executor.go

```diff
@@ -0,0 +1,147 @@
+package rehearse
+
+import (
+	"fmt"
+	"strconv"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	pjapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	pj "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
+	prowconfig "k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type Result struct {
+	Name   string
+	Status pjapi.ProwJobStatus
+}
+
+type Executor struct {
+	dryRun     bool
+	rehearsals []*prowconfig.Presubmit
+	prNumber   int
+	prRepo     string
+	refs       *pjapi.Refs
+	loggers    Loggers
+	pjclient   pj.ProwJobInterface
+	resultChan chan Result
+	errorChan  chan error
+	quit       chan struct{}
+}
+
+func NewExecutor(toBeRehearsed map[string][]prowconfig.Presubmit, prNumber int, prRepo string, refs *pjapi.Refs,
+	loggers Loggers, pjclient pj.ProwJobInterface, dryRun bool, resultChan chan Result, errorChan chan error, quit chan struct{}) *Executor {
+	rehearsals := configureRehearsalJobs(toBeRehearsed, prRepo, prNumber, loggers)
+	return &Executor{
+		dryRun:     dryRun,
+		rehearsals: rehearsals,
+		prNumber:   prNumber,
+		prRepo:     prRepo,
+		refs:       refs,
+		loggers:    loggers,
+		pjclient:   pjclient,
+		resultChan: resultChan,
+		errorChan:  errorChan,
+		quit:       quit,
+	}
+}
+
+// ExecuteJobs takes configs for a set of jobs which should be "rehearsed", and
+// creates the ProwJobs that perform the actual rehearsal. *Rehearsal* means
+// a "trial" execution of a Prow job configuration when the *job config* config
+// is changed, giving feedback to Prow config authors on how the changes of the
+// config would affect the "production" Prow jobs run on the actual target repos
+func (e *Executor) ExecuteJobs() {
+	pjs := e.submitRehearsals()
+	selector, err := e.getExecutionSelector()
+	if err != nil {
+		e.errorChan <- err
+	}
+
+	if !e.dryRun && len(pjs) > 0 {
+		if err := e.waitForJobs(pjs, selector); err != nil {
+			e.errorChan <- err
+			return
+		}
+	}
+
+	e.quit <- struct{}{}
```

Comment:
Close your output channels and block on the receiver end for them to be closed, don't use a `quit` channel